momaland/learning/iql/iql.py

```diff
@@ -0,0 +1,273 @@
+"""Implementation of stateless independent Q-learners. Implemented for the multi-objective congestion game and beach domain."""
+
+import random
+
+import numpy as np
+from gymnasium.spaces import Discrete
+
+from momaland.envs.beach_domain.beach_domain import (
+    MOBeachDomain,
+    _global_capacity_reward,
+    _global_mixture_reward,
+)
+
+
+# from momaland.envs.congestion_game import mocongestion_v0 as CongestionGame
+
+
+class TabularMOBeachDomainWrapper(MOBeachDomain):
+    """Wrapper for the MO-beach domain environment to return only the current beach section as state.
+
+    MO-Beach domain returns 5 observations in each timestep:
+        - agent type
+        - section id
+        - section capacity
+        - section consumption
+        - % of agents of current type
+    In the original paper however tabular Q-learning is used and therefore only the current beach section is used as
+    the state. This provide allows to compare results of the original paper:
+
+    From Mannion, P., Devlin, S., Duggan, J., and Howley, E. (2018). Reward shaping for knowledge-based multi-objective multi-agent reinforcement learning.
+    """
+
+    def __init__(self, **kwargs):
+        """Initialize the wrapper.
+
+        Initialize the wrapper and set the observation space to be a discrete space with the number of sections as
+        possible states.
+
+        Args:
+            **kwargs: keyword arguments for the MO-beach domain environment
+        """
+        self.l_cap_min, self.l_cap_max, self.l_mix_min, self.l_mix_max = kwargs.pop("local_constants")
+        self.g_cap_min, self.g_cap_max, self.g_mix_min, self.g_mix_max = kwargs.pop("global_constants")
+        super().__init__(**kwargs)
+
+        self.observation_spaces = dict(
+            zip(
+                self.agents,
+                [
+                    Discrete(
+                        self.sections,
+                    )
+                ],
+            )
+        )
+
+    def normalize_objective_rewards(self, reward, reward_scheme):
+        """Normalize the rewards based on the reward scheme.
+
+        Args:
+            reward: the reward to normalize
+            reward_scheme: the reward scheme to use
+
+        Returns:
+            np.array: the normalized reward
+        """
+        # Set the normalization constants
+        if reward_scheme == "local":
+            cap_min, cap_max, mix_min, mix_max = self.l_cap_min, self.l_cap_max, self.l_mix_min, self.l_mix_max
+        else:
+            cap_min, cap_max, mix_min, mix_max = self.g_cap_min, self.g_cap_max, self.g_mix_min, self.g_mix_max
+
+        # Normalize the rewards
+        cap_norm = (reward[0] - cap_min) / (cap_max - cap_min)
```

Comment:
I believe we have a NormalizeReward wrapper made for this. Is it different?

Reply:
The NormalizeReward wrapper normalizes the rewards that the agents receive directly. In the BDP environment, agents can be rewarded according to two reward schemes (local/global), but the reported results in the graphs are always using the (normalized) 'global' reward scheme regardless of what the agents receive. This method is primarily used to compute these normalized global rewards.